leakpro/attacks/mia_attacks/yoqo.py

```diff
@@ -0,0 +1,340 @@
+        loss = self.loss(loss, target_output)
+        loss = loss * weights
+        return loss.sum()
+
+    def _optimization_objective_online_binary(
+        self: Self,
+        x0: Tensor,
+        dx: Tensor,
+        target_output: Tensor,
+        weights: Tensor
+    ) -> float:
+        loss = torch.cat([shadow_model.model_obj(x0 + dx) for shadow_model in self.shadow_models], dim = 1)
+        loss = self.loss(loss, target_output)
+        loss = loss * weights
+        return loss.sum()
+
+    def _optimization_objective_offline(
+        self: Self,
+        x0: Tensor,
+        dx: Tensor,
+        target_output: Tensor,
+        weights: Tensor
+    ) -> float:
+        loss = torch.stack([shadow_model.model_obj(x0 + dx) for shadow_model in self.shadow_models], dim = 2)
+        loss = self.loss(loss,target_output)
+        loss = (loss * weights).sum()
+        norm = self.alpha * self.mse(dx, torch.zeros(dx.shape, device = dx.device))
+        return (loss + norm)
+
+    def _optimization_objective_offline_binary(
+        self: Self,
+        x0: Tensor,
+        dx: Tensor,
+        target_output: Tensor,
+        weights: Tensor
+    ) -> float:
+        loss = torch.cat([shadow_model.model_obj(x0 + dx) for shadow_model in self.shadow_models], dim = 1)
+        loss = self.loss(loss,target_output)
+        loss = (loss * weights).sum()
+        norm =  self.alpha * self.mse(dx, torch.zeros(dx.shape, device = dx.device))
+        return (loss + norm)
+
+    def _optimize_xprime(self: Self, x0: Tensor, target_output: Tensor, weights: Tensor) -> Tensor:
+        dx = torch.zeros(x0.shape, device = x0.device)
+        dx.requires_grad = True
+
+        optim = torch.optim.SGD([dx], lr = self.lr_xprime_optimization)
+        for _ in range(self.max_iterations):
+            loss = self._optimization_objective(x0, dx, target_output, weights)
+            if loss < self.stop_criterion:
+                break
+            loss.backward()
+            optim.step()
+            optim.zero_grad()
+
+        return (x0 + dx)
+
+    def run_attack(self:Self) -> CombinedMetricResult:
+        """Runs the attack on the target model and dataset and assess privacy risks or data leakage.
+
+        This method evaluates how the target model's output (logits) for a specific dataset
+        compares to the output of shadow models to determine if the dataset was part of the
+        model's training data or not.
+
+        Returns
+        -------
+        Result(s) of the metric. An object containing the metric results, including predictions,
+        true labels, and signal values.
+
+        """
+
+        data_loader = self.handler.get_dataloader(self.audit_data_indices, batch_size=1)
+
+        predictions = []
+        signal_values = []
+
+        device_name = torch.device("cuda" if torch.cuda.is_available() else "mps" if torch.mps.is_available() else "cpu")
```

Comment:
Not all torch versions have `mps.is_available()`. Perhaps check `torch.has_mps `and if true, look for mps.

Reply:
It's perhaps cleaner to just remove the check? Since there are no checks for MPS in the rest of the code anyway.